Azure Storage Track 2 client libraries

tg-msft opened the pull request.

tg-msft committed:
Azure Storage Track 2 client libraries

kfarmer-msft commented on sdk/storage/.editorconfig:
```diff
@@ -0,0 +1,165 @@
+csharp_style_var_for_built_in_types                                      = true : warning
+csharp_style_var_when_type_is_apparent                                   = true : warning
+csharp_style_var_elsewhere                                               = true : warning
+
```
I'd like to enforce these.

kfarmer-msft replied:
specifically because we don't want it to bifurcate -- picking a single one, I'd pick var.

tg-msft replied:
These are copied as-is from the old repo, so there's no change from what it used to be.  If we want to strengthen, etc., I'd prefer to do that as a separate change because this PR is already incredibly noisy.

kfarmer-msft commented on sdk/storage/Azure.Storage.Blobs/README.md:
````diff
@@ -0,0 +1,115 @@
+
+```Powershell
+Install-Package Azure.Storage.Blobs -Version 1.0.0-preview.1
+```
````
Does markdown have any feature to make a constant out of the version number?

tg-msft replied:
I copied this from the Config service example, but I'll yank the versions because they're not required and then it won't ever need updating.

kfarmer-msft commented on sdk/storage/Swagger/Queues/readme.md:
```diff
@@ -0,0 +1,23 @@
+
+# Generate queue storage
+input-file: ./queue.json
+output-folder: ../../Azure.Storage.Queues/src
```
I'd prefer to have the generated files in a separate folder from non-generated files.  It allows them to be wiped easily as part of normal regeneration, which keeps things clean when removing/renaming things.  Otherwise, they will accumulate in the repository.

tg-msft replied:
I'll add a Generated folder under /src

kfarmer-msft commented on sdk/storage/Swagger/Generator/package-lock.json:
```diff
@@ -0,0 +1,935 @@
+{
```
Should the package-lock files be checked in?

tg-msft replied:
My understanding is that's the best practice

kfarmer-msft commented on sdk/storage/Azure.Storage.Common/src/Azure.Storage.Common.csproj:
```diff
@@ -0,0 +1,21 @@
+      - https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/storage/Azure.Storage.Common/BreakingChanges.txt
+
+      Microsoft Azure Storage team's blog - http://blogs.msdn.com/b/windowsazurestorage/
+    </Description>
```
All this has changed.  See Sean for the new descriptions

seanmcc-msft replied:
https://github.com/wastore/azure-storage-net-pr/commit/1c736cebbbf37a96c0be9e9f36f5c297796e063d

kfarmer-msft replied:
Ted: The quickstarts/tutorials links Sean adjusted may not be relevant (eg, legacy), but the rest should be ok.

tg-msft replied:
Thanks, I'll update.

kfarmer-msft commented on sdk/storage/Azure.Storage.Common/tests/Azure.Storage.Common.Tests.csproj:
```diff
@@ -0,0 +1,14 @@
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.IdentityModel.Clients.ActiveDirectory" Version="4.5.1" />
+    <!-- <ProjectReference Include="..\..\Azure.Storage.Blobs\src\Azure.Storage.Blobs.csproj" /> -->
```
Remove?

tg-msft replied:
Yes, removed.

kfarmer-msft commented on sdk/storage/Azure.Storage.Shared/src/Azure.Storage.Shared.shproj:
```diff
@@ -0,0 +1,15 @@
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory).., Directory.Build.props))\Directory.Build.props" />  
+  <!--
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+  -->
```
Remove?

tg-msft replied:
These files are now gone

kfarmer-msft commented on sdk/storage/Azure.Storage.Shared/tests/Azure.Storage.Tests.Shared.shproj:
```diff
@@ -0,0 +1,15 @@
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory).., Directory.Build.props))\Directory.Build.props" />  
+  <!--
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+  -->
```
Remove?

tg-msft replied:
These files are now gone

kfarmer-msft commented on sdk/storage/Azure.Storage.Shared/tests/Azure.Storage.Tests.Shared.shproj:
```diff
@@ -0,0 +1,15 @@
+<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup Label="Globals">
+    <ProjectGuid>1dfd6176-906a-4620-951b-9ab0ad619fa8</ProjectGuid>
+    <MinimumVisualStudioVersion>14.0</MinimumVisualStudioVersion>
```
Is this a constant for all SDKs?

tg-msft replied:
These files are now gone

kfarmer-msft commented on sdk/storage/Azure.Storage.Shared/tests/TestConfigurationsTemplate.xml:
```diff
@@ -0,0 +1,64 @@
+      <FileServiceEndpoint>http://[ACCOUNT].file.core.windows.net</FileServiceEndpoint>
+      <BlobServiceSecondaryEndpoint>http://[ACCOUNT]-secondary.blob.core.windows.net</BlobServiceSecondaryEndpoint>
+      <QueueServiceSecondaryEndpoint>http://[ACCOUNT]-secondary.queue.core.windows.net</QueueServiceSecondaryEndpoint>
+      <TableServiceSecondaryEndpoint>http://[ACCOUNT]-secondary.table.core.windows.net</TableServiceSecondaryEndpoint>
```
Add one for FileServiceSecondaryEndpoint

kfarmer-msft replied:
and elsewhere

tg-msft replied:
Fixed

kfarmer-msft commented on sdk/storage/Azure.Storage.sln:
```diff
@@ -0,0 +1,131 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio Version 16
```
differs from minver above (14)

FWIW, I've been using VS 2019 (Dev16).  I'm guessing we probably should have no less than VS 2017, as a matter of good NET Standard support, and some features that just don't work well in VS 2015 and earlier.

kfarmer-msft replied:
(like .editorconfig support, which was added directly to 2019)

tg-msft replied:
The minimum version thing was specific to shared projects which are gone.  Our build agent on Windows uses VS 2017.

kfarmer-msft commented on sdk/storage/Azure.Storage.sln:
```diff
@@ -0,0 +1,131 @@
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.28803.352
+MinimumVisualStudioVersion = 10.0.40219.1
```
also differs from above

tg-msft replied:
I think these are fine to leave in because they're standard VS sln properties.

kfarmer-msft commented on sdk/storage/Azure.Storage.sln:
```diff
@@ -0,0 +1,131 @@
+		Release|Any CPU = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(ProjectConfigurationPlatforms) = postSolution
+		{9056EE3A-9ED4-441A-9A88-A567ED07DDDD}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
I recall seeing mention of Preview builds in some of the other projects already in the repo?

tg-msft replied:
I don't know if any of the track 2 libraries have a Preview configuration, but if you check out https://github.com/Azure/azure-sdk-for-net/tree/master/eng/ you can see the work Wes is doing to create a daily feed.

kfarmer-msft commented on sdk/storage/Directory.Build.props:
```diff
@@ -0,0 +1,87 @@
+    across all Storage projects until the need to version independently arises.
+    See /eng/Versioning.props for more details.
+    -->
+    <VersionPrefix>1.0.0</VersionPrefix>
```
So this is the single place we edit our version number?

weshaggard replied:
Yes it should be the single place.

tg-msft replied:
Yes.

weshaggard commented on sdk/storage/.editorconfig:
```diff
@@ -0,0 +1,165 @@
+###############################
```
This should eventually match https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/.editorconfig where we are trying to enforce a common set of standards.

kfarmer-msft replied:
Are we inheriting a config in a higher directory?  I'm guessing not-really, given the fact our sln is rooted lower.  Perhaps an explicit inclusion?

tg-msft replied:
Figuring out how and when to align on these is a thing we should chat about at our next sync.  It might be enough work that it ends up as an item on the roadmap.

weshaggard commented on sdk/storage/Azure.Storage.Blobs/src/Azure.Storage.Blobs.csproj:
```diff
@@ -0,0 +1,25 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <AssemblyTitle>Microsoft Azure.Storage.Blobs client library</AssemblyTitle>
+    <AssemblyName>Azure.Storage.Blobs</AssemblyName>
```
AssemblyName and RootNamespace are redundant and not necessary.

tg-msft replied:
Fixed

weshaggard commented on sdk/storage/Directory.Build.props:
```diff
@@ -0,0 +1,87 @@
+  <!-- DEBUG specific settings -->
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <!-- Treat warnings as errors, except in Debug -->
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
We should probably treat warnings as errors for all configurations.

tg-msft replied:
Yes, I agree.  I made the change, but we can flip it back if folks don't like it.

kfarmer-msft replied:
I'm good with it.

weshaggard commented on sdk/storage/Swagger/Blobs/blob.json:
```diff
@@ -0,0 +1,6974 @@
+{
```
We are trying to keep all the folders directly under `sdk\<service>` as shipping packages so we probably want to consider moving these swagger changes under the corresponding package folders.

tg-msft replied:
I think this ends up being uglier, but I've split everything apart.

weshaggard commented on sdk/storage/Azure.Storage.Shared/src/Azure.Storage.Shared.shproj:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
As we chatted the dotnet cli doesn't support this project type cross-platform, it only works on Windows with a VS install. So I'd suggest just sharing the source files by including them directly in each project.

tg-msft replied:
I've removed these and replaced them with \Shared folders as we discussed.

tg-msft committed:
Update with PR feedback

Removed the Shared projects that were breaking the build and replaced them
with folders that I include/exclude in the .csproj files.

Moved the swagger and generator files out of the top level, moved the generated
code into a directory.

Removed build properties that weren't needed and reordered things to read
a little simpler.

tg-msft committed:
Adding doc warden exclusions and updating links (+ creating targets)

kfarmer-msft commented on sdk/storage/Azure.Storage.Blobs/src/Azure.Storage.Blobs.csproj:
```diff
@@ -1,25 +1,19 @@
-
-      Microsoft Azure Storage team's blog - http://blogs.msdn.com/b/windowsazurestorage/
+      For this release see notes - https://github.com/Azure/azure-storage-net/blob/master/Blob/README.md and https://github.com/Azure/azure-storage-net/blob/master/Blob/Changelog.txt
+      in addition to the breaking changes https://github.com/Azure/azure-storage-net/blob/master/Blob/BreakingChanges.txt
```
The azure-storage-net references should move to the new library.

tg-msft replied:
I updated this out from under you.

kfarmer-msft commented on sdk/storage/Azure.Storage.Blobs/src/Azure.Storage.Blobs.csproj:
```diff
@@ -1,25 +1,19 @@
-      Microsoft Azure Storage team's blog - http://blogs.msdn.com/b/windowsazurestorage/
+      For this release see notes - https://github.com/Azure/azure-storage-net/blob/master/Blob/README.md and https://github.com/Azure/azure-storage-net/blob/master/Blob/Changelog.txt
+      in addition to the breaking changes https://github.com/Azure/azure-storage-net/blob/master/Blob/BreakingChanges.txt
+      Microsoft Azure Storage quickstarts and tutorials - https://docs.microsoft.com/en-us/azure/storage/
```
This is (currently) predicated on the old library.  Perhaps for now point to the local samples project?

tg-msft replied:
I added empty BreakingChanges and ChangeLog files.  Because nothing has technically broken or changed yet.

kfarmer-msft commented on sdk/storage/Azure.Storage.Blobs/src/Azure.Storage.Blobs.csproj:
```diff
@@ -1,25 +1,19 @@
     </Description>
   </PropertyGroup>
   <ItemGroup>
+    <Compile Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Common\src\Shared\**\*.cs" />
```
Add a TO-DO to move this back when the build system can support it?

This is going to make VS a little weird -- it doesn't like to pick up new files like this.

tg-msft replied:
We should dig into this and figure out whether we actually need the Shared files.  Without the pipeline, there's so much less code being shared that we might be able to find it a single home somewhere.

kfarmer-msft replied:
Worth discussing.  There are some shared concepts like SAS generation that have slight service-specific differences that need to be catered to, which probably should *not* lead to types that can be shared across services, but which are complex enough that you want to keep from having to maintain duplicate code.  

Unlike the Common library, which is for things which can/should be shareable.

tg-msft committed:
Add the service version to the READMEs and a doc-warden exclusion

weshaggard commented:
https://dev.azure.com/azure-sdk/public/_build/results?buildId=30661
```
/opt/hostedtoolcache/dncs/2.1.503/x64/sdk/2.1.503/Microsoft.Common.CurrentVersion.targets(4551,5): error MSB3030: Could not copy the file "/home/vsts/work/1/s/sdk/storage/Azure.Storage.Common/tests/Shared/TestConfigurations.xml" because it was not found. [/home/vsts/work/1/s/sdk/storage/Azure.Storage.Blobs/tests/Azure.Storage.Blobs.Tests.csproj]
```

Looks like you are missing this TestConfigurations.xml file but only on Linux. Where is that file supposed to come from?

tg-msft commented:
Yes, I'm not sure why it didn't appear.  It's automatically created at https://github.com/Azure/azure-sdk-for-net/blob/d0c2b50da7059a28b7ceb2e9d046eaf901bd6da6/sdk/storage/Azure.Storage.Common/tests/Azure.Storage.Common.Tests.csproj#L12

weshaggard commented:
It could be an ordering thing? What is it that makes Azure.Storage.Common the first project that gets built?Is there any reason we don't just check-in this file?

tg-msft commented:
Don't the project references (Blobs.Tests referencing Common.Tests) cause them to be ordered correctly?  And we don't want to check this file in, for reasons.

weshaggard commented:
>Don't the project references (Blobs.Tests referencing Common.Tests) cause them to be ordered correctly? And we don't want to check this file in, for reasons.

Perhaps I'm missing it but I don't see Blobs.Tests doing a ProjectReference to Common.Tests project.

tg-msft commented:
That's...  a good point.  We were referencing the Shared project but not Common.  Let me try working around this by copying *.xml because the test code is resilient enough to not care.

tg-msft committed:
Be less aggressive about which XML files get copied for testing

as they're never relevant to server builds

weshaggard commented on .gitignore:
```diff
@@ -115,6 +115,7 @@ tools/*.dll
 # Sensitive files
 *.keys
 *.pfx
+sdk/storage/Azure.Storage.Common/tests/Shared/TestConfigurations.xml
```
As something to clean-up, in general we don't like to generate files into the source tree during the build. This file should either be committed or if it needs to be generated then it should get generated under the intermediate directory.

kfarmer-msft commented on sdk/storage/Azure.Storage.Blobs/tests/Azure.Storage.Blobs.Tests.csproj:
```diff
@@ -10,7 +10,7 @@
   <ItemGroup>
     <Compile Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Common\tests\Shared\**\*.cs" />
-    <None Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Common\tests\Shared\TestConfigurations.xml">
+    <None Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Common\tests\Shared\*.xml">
```
Motivation?

tg-msft replied:
There's no explicit ordering so the file may not have been created by the time this project builds.  (And we don't care at that point - if you don't have a config file, it doesn't matter if we copy it.)

kfarmer-msft commented:
> It could be an ordering thing? What is it that makes Azure.Storage.Common.Tests the first project that gets built?Is there any reason we don't just check-in this file?

TestConfigurations.xml holds credentials for testing, and must *NOT* be checked in.  Keeping it excluded from Git means there won't be any accidental leaks via this file.

weshaggard commented:
>TestConfigurations.xml holds credentials for testing, and must NOT be checked in. Keeping it excluded from Git means there won't be any accidental leaks via this file.

I'd suggest that we generate it as a pre-test step in each of the test projects then. I assume you pull in the secrets as environment variables from keyvault during the build setup, in which case I'd simply have msbuild write out the file based on those secrets.

kfarmer-msft commented:
> > TestConfigurations.xml holds credentials for testing, and must NOT be checked in. Keeping it excluded from Git means there won't be any accidental leaks via this file.
> 
> I'd suggest that we generate it as a pre-test step in each of the test projects then. I assume you pull in the secrets as environment variables from keyvault during the build setup, in which case I'd simply have msbuild write out the file based on those secrets.

Bad assumption :)

Typically, the credentials have to be created on the spot or handed around from server devs to us, since they're accounts for features-in-development.  Not stable enough to end up in KeyVault.

tg-msft committed:
Adding Live test categtory to exclude for the full framework

weshaggard commented on sdk/storage/Azure.Storage.Common/tests/Shared/AssemblyInfo.cs:
```diff
@@ -0,0 +1,8 @@
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+[assembly: Parallelize(Workers = 0, Scope = ExecutionScope.MethodLevel)]
```
Comment this attribute out to disable the parallel test running which should hopeful get your tests passing until we can dig in and fix the real issues.

tg-msft committed:
Disabling test parallelizaton temporarily
